### PR TITLE
Wire up custom event loop interop for the GLFW embedder.

### PR DESCRIPTION
**Pull request by chinmaygarde**

### CODE REVIEW

**Comment by chinmaygarde** [comment]
Fixes https://github.com/flutter/flutter/issues/30730

File: tools/gn [hunk 1]
```diff
@@ -315,6 +316,9 @@ def parse_args(args):
                       help='On the mac, the SDK is inferred from the Xcode location unless this flag is specified. ' +
                       ' Setting the FLUTTER_MAC_SDK_PATH environment variable achieves the same effect.')
 
+  parser.add_argument('--build-glfw-shell', dest='build_glfw_shell', default=False, action='store_true',
+                      help='Force building the GLFW shell on platforms where it is not built by default.')
```
Comment (stuartmorgan) [inline tools/gn]: s/platforms/desktop platforms/?

**Comment by wjljack** [comment]
when to fix windows cpu too high bug?

**Comment by chinmaygarde** [comment]
Ok. This is good to go.

File: shell/platform/glfw/BUILD.gn [hunk 2]
```diff
@@ -65,6 +67,11 @@ source_set("flutter_glfw") {
       "$flutter_root/shell/platform/linux/config:gtk3",
       "$flutter_root/shell/platform/linux/config:x11",
     ]
+  } else if (is_mac) {
```
Comment (stuartmorgan) [inline shell/platform/glfw/BUILD.gn]: This seems unrelated to the PR?

File: shell/platform/glfw/glfw_event_loop.cc [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+bool GLFWEventLoop::RunsTasksOnCurrentThread() const {
+  return std::this_thread::get_id() == main_thread_id_;
+}
+
+void GLFWEventLoop::WaitForEvents() {
+  const auto now = TaskTimePoint::clock::now();
+  std::vector<FlutterTask> expired_tasks;
+
+  // Process exipred tasks.
```
[the remark this answers is not in the document]
Comment (stuartmorgan) [reply shell/platform/glfw/glfw_event_loop.cc]: The function is doing three distinct things, and is 50+ lines long, so by common evaluations of whether a function is small, there's a pretty good argument that this is already "larger".

I'm not clear what advantage there is in not pulling out at least the two non-trivial, self-contained blocks into helpers. I'm not going to hold up the PR on this, but I'd like to understand the argument for not breaking this up.

File: shell/platform/glfw/glfw_event_loop.cc [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+      // done processing expired tasks.
+      expired_tasks.push_back(task_queue_.top().task);
+
+      // Remove the tasks from the delayed tasks queue.
+      task_queue_.pop();
+    }
+  }
+
+  // Fire expired tasks.
+  {
```
Comment (stuartmorgan) [inline shell/platform/glfw/glfw_event_loop.cc]: There's no scoped lock; why is there a scope block? (I would argue that the fact that you did this despite in not being any mechanical reason I can see is an indication that it should be broken into helpers.)

File: shell/platform/glfw/glfw_event_loop.cc [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    // Flushing tasks here without holing onto the task queue mutex.
+    for (const auto& task : expired_tasks) {
+      on_task_expired_(&task);
+    }
+  }
+
+  // Sleep till the next task needs to be processed. If a new task comes
+  // along, the wait in GLFW will be resolved early because we posted an empty
```
Comment (stuartmorgan) [inline shell/platform/glfw/glfw_event_loop.cc]: s/we posted/PostTask posts/

(go/avoidwe)

File: shell/platform/glfw/glfw_event_loop.cc [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    uint64_t flutter_target_time_nanos) {
+  const auto now = TaskTimePoint::clock::now();
+  const auto flutter_duration =
+      flutter_target_time_nanos - FlutterEngineGetCurrentTime();
+  return now + std::chrono::nanoseconds(flutter_duration);
+}
+
+void GLFWEventLoop::PostTask(FlutterTask flutter_task,
+                             uint64_t flutter_target_time_nanos) {
+  static std::atomic_uint64_t gGlobalTaskOrder(0);
```
[the remark this answers is not in the document]
Comment (stuartmorgan) [reply shell/platform/glfw/glfw_event_loop.cc]: `g` comes from Hungarian notation and means "global". This doesn't have global scope, and I have a strong opinion that annotating something that's not a global as if it were is wrong :) Please use `s`, as I would find `g` actively confusing.

File: shell/platform/glfw/glfw_event_loop.cc (hunk 5 above)
Comment (chinmaygarde) [reply shell/platform/glfw/glfw_event_loop.cc]: Done.

File: shell/platform/glfw/glfw_event_loop.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (chinmaygarde) [reply shell/platform/glfw/glfw_event_loop.cc]: Fair point. Updated. Will patch the other uses in the engine a separate patch.
